File: java源码/offer/Test18.java
package offer;

/**
 * 剑指offer-题18：树的子结构
 * 
 * 输入两棵二叉树A，B，判断B是不是A的子结构。（ps：我们约定空树不是任意一个树的子结构）
 */

public class Test18 {
	public boolean HasSubtree(TreeNode root1, TreeNode root2) {
		boolean result = false;
		if (root1 != null && root2 != null) {
			if (root1.val == root2.val) {
				result = DoseTree1HavaTree2(root1, root2);
			}
			if (!result) {
				result = HasSubtree(root1.left, root2);
			}
			if (!result) {
				result = HasSubtree(root1.right, root2);
			}
		}
		return result;
	}

	public boolean DoseTree1HavaTree2(TreeNode root1, TreeNode root2) {
		if (root2 == null) {
			return true;
		}
		if (root1 == null) {
			return false;
		}
		if (root1.val != root2.val) {
			return false;
		}
		return DoseTree1HavaTree2(root1.left, root2.left) && DoseTree1HavaTree2(root1.right, root2.right);
	}
	
	
	public static void main(String[] args) {
		TreeNode root = new TreeNode(1);
        TreeNode a = new TreeNode(2);
        TreeNode b = new TreeNode(3);
        TreeNode c = new TreeNode(4);
        TreeNode d = new TreeNode(5);
        TreeNode e = new TreeNode(6);
        TreeNode f = new TreeNode(7);
        TreeNode g = new TreeNode(8);
        root.left = a;
        root.right = b;
        a.left = c;
        c.right = f;
        b.left = d;
        b.right = e;
        e.left = g;
        
        TreeNode root2 = new TreeNode(6);
		TreeNode a2 = new TreeNode(8);
		root2.left = a2;
//		TreeNode b2 = new TreeNode(9);
//		root2.right = b2;
		System.out.println(new Test18().HasSubtree(root, root2));
		
	}
}

File: python3源码/Test18.py
# -*- coding:utf-8 -*-

# 剑指offer-题18：树的子结构

# 输入两棵二叉树A，B，判断B是不是A的子结构。（ps：我们约定空树不是任意一个树的子结构）


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def HasSubtree(self, pRoot1, pRoot2):
        result = False
        if pRoot1 is not None and pRoot2 is not None:
            if pRoot1.val == pRoot2.val:
                result = self.DoseTree1HaveTree2(pRoot1, pRoot2)
            if not result:
                result = self.HasSubtree(pRoot1.left, pRoot2)
            if not result:
                result = self.HasSubtree(pRoot1.right, pRoot2)
        return result

    def DoseTree1HaveTree2(self, pRoot1, pRoot2):
        if pRoot2 is None:
            return True
        if pRoot1 is None:
            return False
        if pRoot2.val != pRoot1.val:
            return False
        return self.DoseTree1HaveTree2(pRoot1.right, pRoot2.right) and self.DoseTree1HaveTree2(pRoot1.left, pRoot2.left)


solution = Solution()
root = TreeNode(1)
a = TreeNode(2)
b = TreeNode(3)
c = TreeNode(4)
d = TreeNode(5)
e = TreeNode(6)
f = TreeNode(7)
g = TreeNode(8)
root.left = a
root.right = b
a.left = c
c.right = f
b.left = d
b.right = e
e.left = g

root2 = TreeNode(6)
a2 = TreeNode(8)
root2.left = a2
print(solution.HasSubtree(root, root2))
